src/hooks/UrqlUseSubscriptionWithHandler.re

```diff
@@ -48,25 +52,28 @@ let useSubscriptionResponseToRecordWithHandler =
   {fetching, data, error, response};
 };
 
-let useSubscriptionWithHandler =
+let useSubscription =
     (
-      ~request: UrqlTypes.request('response),
-      ~handler: UrqlTypes.parsedHandler('acc, 'response),
-    ) => {
+      type a, type b, type c,
+      ~handler: handler(a, b, c),
+      request: UrqlTypes.request(b),
+    ): useSubscriptionResponse(c) => {
   let parse = request##parse;
-  let parsedHandler =
-      (~prevSubscriptions: option('acc), ~subscription: Js.Json.t) =>
-    handler(~prevSubscriptions, ~subscription=parse(subscription));
-
   let args =
-    useSubscriptionWithHandlerArgs(
+    useSubscriptionArgs(
       ~query=request##query,
       ~variables=request##variables,
       (),
     );
 
-  let state = useSubscriptionWithHandlerJs(args, parsedHandler)[0];
-  let useSubscriptionResponse =
-    state |> useSubscriptionResponseToRecordWithHandler;
-  useSubscriptionResponse;
-};
\ No newline at end of file
+  let state: useSubscriptionResponse(c) = switch (handler) {
+    | NoHandler => 
+      useSubscriptionJs(args, None)[0] 
+      |> useSubscriptionResponseToRecord(parse)
+    | Handler(handler_fn) =>
+      useSubscriptionJs(args, Some(handler_fn))[0] 
```

Comment:
So one thing I'm concerned about here – from what I can understand, the individual subscriptions coming in as the second argument to `handler` _won't_ get parsed properly in this implementation, which would break type inference. That was the idea behind `parsedHandler`. Basically, we would apply the user supplied `handler` to `prevSubscriptions` and `parse(subscription)`, such that each new `subscription` coming in would have its type inferred. Does this do the same? I.e. if I passed this request:

```
module SubscribeRandomInt = [%graphql
  {|
  subscription subscribeNumbers {
    newNumber @bsDecoder(fn: "string_of_int")
  }
|}
];
```

with this `handler`:

```
let handler = (~prevSubscriptions, ~subscription) => {
  switch (prevSubscriptions) {
  | Some(subs) => Array.append(subs, [|subscription|])
  | None => [|subscription|]
  };
};
```

I'd expect the type of `data` to be inferred properly as `array({. "newNumber": string })`. Does that happen in the example w/ this implementation?

Reply:
Nice catch, it really isn't properly parsing the second argument before supplying it to handler. It wouldn't mess up type inference, but it does mess up type safety.